scanner.cs

```diff
@@ -3,66 +3,176 @@
 
-	private int fsaDigit() {
-		return 0; // change this
-	}
+    private Tuple<string, TOKENS> fsaDigit() {
+        string lexeme = "";
+        string DIGITS = Constants.DIGITS;
+        TOKENS token;
+        char next;
+        goto S0; // Smothers warning about S0 label not used
+        S0: // Start state
+            next = __bytes[__curByte];
+            __column++;
+            __curByte++;
+            if(DIGITS.Contains("" + next)){
+                lexeme += next;
+                goto S1;
+            } else {
+                __curByte--;
+                throw new Exception(
+                    "Dispatcher done goofed. Passed " +
+                    next + " to the digit FSA, but " +  next +
+                    "is not a digit."
+                );
+            }
+        S1: // One or more digits have been read
+            token = TOKENS.INTEGER_LIT;
+            next = __bytes[__curByte];
+            __column++;
+            __curByte++;
+            if(DIGITS.Contains("" + next)){
+                lexeme += next;
+                goto S1;
+            } else if(next == '.'){
+                lexeme += next;
+                goto S2;
+            } else if(next == 'e' || next == 'E'){
+                lexeme += next;
+                goto S4;
+            } else { //This is a success state, so reset the fp and return the lexeme
+                __curByte--; //reset the fp
+                return new Tuple<string, TOKENS>(lexeme, token);
+            }
+        S2: // A '.' has been read
+            next = __bytes[__curByte];
+            __column++;
+            __curByte++;
+            if(DIGITS.Contains("" + next)){
+                lexeme += next;
+                goto S3;
+            } else {
+                // Must remove the last character (.) from lexeme
+                lexeme.Remove(lexeme.Length - 1);
+                __curByte -= 2;
+                return new Tuple<string, TOKENS>(lexeme, token);
+            }
+        S3: // Digits have followed a valid '.'
+            token = TOKENS.FIXED_LIT;
+            next = __bytes[__curByte];
+            __column++;
+            __curByte++;
+            if(DIGITS.Contains("" + next)){
+                lexeme += next;
+                goto S3;
+            } else if(next == 'e' || next == 'E'){
+                lexeme += next;
+                goto S4;
+            } else {
+                __curByte--;
+                return new Tuple<string, TOKENS>(lexeme, token);
+            }
+        S4: // An e or E has been read
+            next = __bytes[__curByte];
+            __column++;
+            __curByte++;
+            if(next == '+' || next == '-'){
+                lexeme += next;
+                goto S5;
+            } else {
+                // Must remove the last character (e or E) from lexeme
+                lexeme = lexeme.Remove(lexeme.Length - 1);
+                __curByte -= 2;
+                return new Tuple<string, TOKENS>(lexeme, token);
```

Comment:
@JesseStephenBrown I understand. Until I test though, I am worried this will catch invalid tokens.. Like "1234e1234" I strongly recommend some automated tests to run through these FSA's before we get too far into this.